### PR TITLE
 tutorial/handling null values in protobuf and avro

### PR DESCRIPTION
**Pull request by Cerchie**
need to figure out a smooth way to handle testing and a way to teach the reader about what happens when null values are handled/not. tagging @davetroiano

### CODE REVIEW

File: handling-null-values-in-avro-and-protobuf/kafka/.gitignore [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# Default ignored files
```
Comment (davetroiano) [inline handling-null-values-in-avro-and-protobuf/kafka/.gitignore:1]: since everything in here is petty generic (gradle, intellij, etc), you can add to the root level `.gitignore` as needed and delete this file

File: handling-null-values-in-avro-and-protobuf/kafka/README.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
+In this tutorial let's say we're tracking purchase events in Kafka with Confluent Cloud, each with an `item`, a `total_cost`, and a `customer_id`.
```
Comment (davetroiano) [inline handling-null-values-in-avro-and-protobuf/kafka/README.md:1]: add a title, something like

```
# How to allow `null` field values in Avro and Protobuf
```

File: handling-null-values-in-avro-and-protobuf/kafka/README.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+Login, and then click 'Environments -> Create Cloud Environment' and create a cloud environment using the defaults there. 
+
+Navigate to your environment and click 'Add cluster'. Create a cluster using the default values provided. 
+
+Click 'Topics -> Add topic' to create two topics with the default values, one named 'avro-purchase' and the other 'proto-purchase' (we'll cover null values in Protobuf schemas later in the tutorial). 
+
+On the right-hand navbar, click 'API keys -> Add key -> Global access'. Download the values as you will need them to run this tutorial. 
+
+In the same navbar, click 'Clients -> Choose Your Language -> Java -> Create Schema Registry API key'. Save this key and secret as well as the URL listed in the configuration snippet.
```
Comment (davetroiano) [inline handling-null-values-in-avro-and-protobuf/kafka/README.md:15]: keep the tutorial agnostic wrt where Kafka is running... suggest adding a section to the bottom like [this](https://github.com/confluentinc/tutorials/blob/master/aggregating-minmax/flinksql/README.md?plain=1#L33) though maybe just with 2 sections for now, one to run the unit test, one to run in Confluent Cloud. the third would be `Local in Docker` if you want to tackle that or that can come later

File: handling-null-values-in-avro-and-protobuf/kafka/README.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+On the right-hand navbar, click 'API keys -> Add key -> Global access'. Download the values as you will need them to run this tutorial. 
+
+In the same navbar, click 'Clients -> Choose Your Language -> Java -> Create Schema Registry API key'. Save this key and secret as well as the URL listed in the configuration snippet. 
+
+Create a folder anywhere you like to house the code this tutorial:
+
+```bach
+mkdir handling-null-values
+```
```
Comment (davetroiano) [inline handling-null-values-in-avro-and-protobuf/kafka/README.md:21]: shouldn't need to do this or the copy/paste instructions below to populate files. instead instruct the user to clone the repo if they haven't already (e.g., see [here](https://github.com/confluentinc/tutorials/blob/master/kafka-producer-application/kafka/README.md?plain=1#L74-L78))

File: handling-null-values-in-avro-and-protobuf/kafka/README.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+For the "item" field in `purchase.avsc`, put:
+
+```
+{
+  "type":"record",
+  "namespace": "io.confluent.developer.avro",
+  "name":"PurchaseAvro",
+  "fields": [
+    {"name": "item", "type": ["string", "null"] },
```
Comment (davetroiano) [inline handling-null-values-in-avro-and-protobuf/kafka/README.md:151]: I think this is the first place that I see how to allow null values with Avro... I think that this should be presented much earlier on so that you get the answer even if you don't run the example. e.g., the tutorial could start with something like:

Suppose you are producing Avro- or Protobuf-formatted events to Kafka and you want to allow for a field to be `null`. For example, \<insert case where this makes sense>.

Avro natively supports nullable fields via the `"null"` type. In the \<insert example> example, in order to make the \<field name> field nullable, you can simply allow the type to be `"string"` or `"null"`:

```noformat
{"name": "item", "type": ["string", "null"] }
```

On the other hand, handling null in Protobuf-formatted events is trickier because \<go on with explanation and example>

File: handling-null-values-in-avro-and-protobuf/kafka/README.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
+In this tutorial let's say we're tracking purchase events in Kafka with Confluent Cloud, each with an `item`, a `total_cost`, and a `customer_id`. 
+
+Let's say you're using an avro schema, and sometimes you want to set the `item` to null. How to adjust the schema to allow a null value? Let's get started.
```
Comment (davetroiano) [inline handling-null-values-in-avro-and-protobuf/kafka/README.md:3]: suggest explaining why `item` might be null (... "sometimes you want to set `item` to null, e.g., if xyz") so that it feels a little more realistic. or, if null item is too much of a stretch, pick a different field where null would make sense. perhaps have a `store_id` field that gets set to null if a purchase is made online

File: handling-null-values-in-avro-and-protobuf/kafka/README.md [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+```
+
+Look at `ProtoProducerApp.java`, lines 76-77:
+
+```java
+        purchaseBuilder.setCustomerId("Customer Null")
+                .setTotalCost(random.nextDouble() * random.nextInt(100));
+``` 
+
+We can see that the developer who wrote this app 'forgot' to write the `setItem()` method that adds an item. This means that the value will be null. But when you run you run `./gradlew runProtoProducer` and `./gradlew runProtoConsumer` no errors will arise. That's because Protobuf automatically handles default values.
```
Comment (davetroiano) [inline handling-null-values-in-avro-and-protobuf/kafka/README.md:185]: what makes it into Kafka in this case? empty string? suggest saying that here maybe with a link to protobuf doc about default values per type

File: handling-null-values-in-avro-and-protobuf/kafka/README.md [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+Now, if you _explicitly_ set the value of the item to null:
+
+
+```java
+        purchaseBuilder.setCustomerId("Customer Null").setItem(null)
+                .setTotalCost(random.nextDouble() * random.nextInt(100));
+``` 
+
+In this case, you'll receive a NullPointer error.
```
Comment (davetroiano) [inline handling-null-values-in-avro-and-protobuf/kafka/README.md:195]: whats going to come next? the field being `oneof` where one is a boolean to signify null?

File: handling-null-values-in-avro-and-protobuf/kafka/build.gradle [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+    implementation 'org.apache.logging.log4j:log4j-api:2.18.0'
+    implementation 'org.apache.logging.log4j:log4j-core:2.18.0'
+    implementation 'org.apache.logging.log4j:log4j-slf4j-impl:2.18.0'
+
+    implementation('org.apache.kafka:kafka-streams:3.2.1') {
+        exclude group: 'org.apache.kafka', module: 'kafka-clients'
+    }
+    implementation('org.apache.kafka:kafka-clients:3.2.1')
+    testImplementation 'junit:junit:4.13.1'
```
Comment (davetroiano) [inline handling-null-values-in-avro-and-protobuf/kafka/build.gradle:54]: suggest more recent dependency versions so that this is fresher... maybe `7.5.1` for confluent dependencies and `3.6.1` for AK

File: handling-null-values-in-avro-and-protobuf/kafka/build.gradle [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    id 'java'
+    id 'idea'
+    id 'eclipse'
+    id "com.google.protobuf" version "0.9.4"
+    id "com.github.imflog.kafka-schema-registry-gradle-plugin" version "1.13.0"
+    id "com.github.davidmc24.gradle.plugin.avro" version "1.3.0"
+}
+
+sourceCompatibility = JavaVersion.VERSION_11
+targetCompatibility = JavaVersion.VERSION_11
```
Comment (davetroiano) [inline handling-null-values-in-avro-and-protobuf/kafka/build.gradle:26]: for consistency with the rest of the repo

```suggestion
java {
    sourceCompatibility = JavaVersion.VERSION_17
    targetCompatibility = JavaVersion.VERSION_17
}
```

**Review by davetroiano — COMMENTED** [review]
LG so far @Cerchie!

File: handling-null-values-in-avro-and-protobuf/kafka/README.md (hunk 4 above)
Comment (davetroiano) [reply handling-null-values-in-avro-and-protobuf/kafka/README.md]: exactly - if you add this to the top-level `settings.gradle`:

```
include 'handling-null-values-in-avro-and-protobuf:kafka'
```

Then you'd be able to run it from the top level:

```
./gradlew clean :handling-null-values-in-avro-and-protobuf:kafka:test --info  
```

File: handling-null-values-in-avro-and-protobuf/kafka/README.md (hunk 6 above)
Comment (davetroiano) [reply handling-null-values-in-avro-and-protobuf/kafka/README.md]: looks good with the donation case!

File: handling-null-values-in-avro-and-protobuf/kafka/README.md (hunk 8 above)
Comment (davetroiano) [reply handling-null-values-in-avro-and-protobuf/kafka/README.md]: Does `optional` do the trick? i.e. does everything just work if you do that? I'm looking at the guidance here:

>If not using optional would add complexity or ambiguity, then use optional primitive fields. Wrapper types must not be used going forward.

which makes me think we might want to avoid the wrapper pattern... I'm not sure why this is the guidance, e.g., will wrapper types eventually be deprecated?

I'm thinking that if `optional` works, go with that, and we'd also want to update [this](https://docs.confluent.io/cloud/current/sr/fundamentals/serdes-develop/serdes-protobuf.html#handling-null-values-with-wrappers) docs section to recommend `optional` rather than use a wrapper

wdyt?

File: handling-null-values-in-avro-and-protobuf/kafka/README.md [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+  "name":"PurchaseAvro",
+  "fields": [
+    {"name": "item", "type": ["string", "null"] },
+    {"name": "total_cost", "type": "double" },
+    {"name": "customer_id", "type": "string"}
+  ]
+}
+```
+
+When you run `gradle runAvroProducer` and furthermore, `gradle runAvroConsumer`, you'll see that the events with null items are produced and consumed successfully.
```
Comment (davetroiano) [inline handling-null-values-in-avro-and-protobuf/kafka/README.md:164]: use the top level gradle wrapper so gradle doesnt have to be a dependency. from the top level I think it'd be:

```
./gradlew :handling-null-values-in-avro-and-protobuf:kafka:runAvroProducer
```

the same applies to the `gradle` commands below for the protobuf example

File: handling-null-values-in-avro-and-protobuf/kafka/README.md (hunk 4 above)
Comment (Cerchie) [reply handling-null-values-in-avro-and-protobuf/kafka/README.md]: ok replacing

File: handling-null-values-in-avro-and-protobuf/kafka/README.md (hunk 8 above)
Comment (Cerchie) [reply handling-null-values-in-avro-and-protobuf/kafka/README.md]: so with this protobuf definition:

```
message Purchase {
  optional string item = 1;
  double total_cost = 2;
  string customer_id = 3;
}
```

you still end up with a `NullPointerException` ¯\_(ツ)_/¯

but that's a good point about the guidance, I'll pass it on to docs.

File: handling-null-values-in-avro-and-protobuf/kafka/src/main/proto/purchase.proto [hunk 12]
```diff
@@ -0,0 +1,10 @@
+syntax = "proto3";
+
+package io.confluent.developer.proto;
+option java_outer_classname = "PurchaseProto";
```
[the remark this answers is not in the document]
Comment (Cerchie) [reply handling-null-values-in-avro-and-protobuf/kafka/src/main/proto/purchase.proto:4]: so weird -- I updated with that option and then ran `gradle clean build` and then I got all these errors. `Purchase` wasn't recognized, and `PurchaseProto` didn't have a `Builder`. is there a different command to use?

[the remark this answers is not in the document]
Comment (davetroiano) [reply handling-null-values-in-avro-and-protobuf/kafka/src/main/proto/purchase.proto:4]: Did you update the imports?

[the remark this answers is not in the document]
Comment (Cerchie) [reply handling-null-values-in-avro-and-protobuf/kafka/src/main/proto/purchase.proto:4]: yes!

[the remark this answers is not in the document]
Comment (Cerchie) [reply handling-null-values-in-avro-and-protobuf/kafka/src/main/proto/purchase.proto:4]: works better now

**Comment by davetroiano** [comment]
closing after discussion with @Cerchie
